volumetric/volMask.cpp

```diff
@@ -0,0 +1,255 @@
+        if( *pIt > *maxIt ){*maxIt = *pIt;}
+        minIt++;
+        maxIt++;
+      }
+    }
+  }
+  
+  // offset to avoid problems on borders
+  Z3i::Point offset(5,5,5);
+  minP -= offset;
+  maxP += offset;
+  trace.info() << "sub-domain:" << minP << " " << maxP << std::endl;
+  return typename TImageMask::Domain(minP, maxP);
+}
+
+
+template<typename TImage, typename TImageMask>
+void
+applyMask(const TImage &inputImage,TImage &outputImage,
+          const TImageMask &maskImage, typename TImageMask::Value maskValue)
+{
+  for (const auto &p: outputImage.domain())
+  {
+    if (inputImage.domain().isInside(p) &&  maskImage(p) ==  maskValue)
+    {
+      outputImage.setValue(p, inputImage(p) );
+    }
+  }
+  
+}
+
+template<typename TImage, typename TImageMask>
+void
+processImage(const TImage &inputImage, const TImageMask &maskImage,
+             typename TImageMask::Value maskValue, std::string outputFileName ){
+  // First step getting the bounding box of the domain:
+  auto subDm = subDomainMasked(inputImage, maskImage, maskValue);
+  TImage outputImage( subDm );
+  // Second step: masking source image
+  applyMask(inputImage, outputImage, maskImage,  maskValue);
+  trace.info() << "writing output image...";
+  GenericWriter<TImage>::exportFile(outputFileName, outputImage);
+}
+
+
+int main( int argc, char** argv )
+{
+  // parse command line -------------------------------------------------------
+  po::options_description general_opt("Allowed options are");
+  general_opt.add_options()
+  ("help,h", "display this message")
+#ifdef WITH_ITK
+  ("input,i", po::value<std::string >(), "an input vol (or ITK: .nii, mha, ... ) file. " )
+  ("inputType,t", po::value<std::string>()->default_value(""), "to sepcify the input image type (int or double)." )
+#else 
+  ("input,i", po::value<std::string >(), "an input vol file. " )
+#endif
+  ("mask,a", po::value<std::string >(), "the mask image that represents the elements that are copied as output in the resulting image (by default set to 1 you can change this value by using --maskValue).  " )
+  ("output,o", po::value<std::string >()->default_value("result.vol"), "the output masked image." )
+  ("maskValue,m", po::value<int>()->default_value(1), "the masking value." );
+  
+  
+  bool parseOK=true;
+  po::variables_map vm;
+  try
+  {
+    po::store(po::parse_command_line(argc, argv, general_opt), vm);
+  }
+  catch(const std::exception& ex)
+  {
+    parseOK=false;
+    trace.info()<< "Error checking program options: "<< ex.what()<< endl;
+  }
+  
+  
+  // check if min arguments are given and tools description ------------------
+  po::notify(vm);
+  if( !parseOK || vm.count("help")||argc<=1)
+  {
+    std::cout << "Usage: " << argv[0] << " [input]\n"
+    << "Extracts a new image from the a mask image that represents the regions of the image which are selected and copied in the resulting image. Elements outside the regions defined by the mask are set to 0.\n"
```

Comment:
"the a" ?
In fact, it is more : "Outputs a new image from two input images with same size, one representing the data, one representing the selection mask. The size of output image is the size of the bounding box of selected values, plus the chosen border offset ."

Reply:
Question: do you check that input and mask image have compatible sizes ? Or is it not necessary ?

Reply:
Thanks for the suggestion.
No check are done, with the last commit a warning will mention that no masked values are found and the resulting image will be empty (image with 0 and domain of source image)